shared/middleware/files_api.rb

```diff
@@ -805,6 +813,12 @@ def files_put_file(encrypted_channel_id, filename, body)
     file
   end
 
+  def project_type(encrypted_channel_id, storage_apps)
+    storage_apps.get(encrypted_channel_id)[:projectType]
+  rescue
```

Comment:
Oof. I thought I was good to go here because I'm using [`StorageApps.get`](https://github.com/code-dot-org/code-dot-org/blob/1edbf48dc7c47740539bba065e086f99e57bc8fe/shared/middleware/helpers/storage_apps.rb#L47-L67) which contains this logic accounting for the very thing you've described:

https://github.com/code-dot-org/code-dot-org/blob/1edbf48dc7c47740539bba065e086f99e57bc8fe/shared/middleware/helpers/storage_apps.rb#L57-L59

Unfortunately it looks like the value extracted there doesn't end up in the merged row returned by the `get` method, so you are correct, this is currently broken.

Reply:
ok, maybe this is what you are saying, but it looks like value['projectType'] is only ever set from the `projectType` row in the `storage_apps` table https://github.com/code-dot-org/code-dot-org/blob/1edbf48dc7c47740539bba065e086f99e57bc8fe/shared/middleware/helpers/storage_apps.rb#L247

 so I agree with your conclusion that the `projectType` json field does not incorporate the logic of checking for the `level` json field, although I have not confirmed it empirically.

Reply:
It looks like, when it is present, the `"level"` field in the `value` JSON gives a consistent representation of the project type:

```
mysql> select SUBSTRING(level_part, 1, INSTR(level_part, '"')-1) as level, count(id) from (select id, SUBSTRING(value, INSTR(value, '"level":"')+9) as level_part from storage_apps where id > 52387852 and project_type is null and INSTR(value, '"level":"')) as t1 group by level order by count(id) desc;
+--------------------------------+-----------+
| level                          | count(id) |
+--------------------------------+-----------+
| /projects/gamelab              |      5174 |
| /projects/applab               |      4912 |
| /projects/weblab               |       545 |
| /projects/artist               |       466 |
| /projects/frozen               |       458 |
| /projects/playlab              |       193 |
| /projects/playlab_k1           |       156 |
| /projects/flappy               |       128 |
| /projects/starwarsblocks_hour  |        10 |
| /projects/minecraft_hero       |         7 |
| /projects/algebra_game         |         6 |
| /projects/minecraft_designer   |         4 |
| /projects/bounce               |         3 |
| /projects/minecraft_adventurer |         3 |
| /projects/artist_k1            |         3 |
| /projects/eval                 |         1 |
| /projects/iceage               |         1 |
| /projects/gumball              |         1 |
| /projects/basketball           |         1 |
+--------------------------------+-----------+
```

However, there's a significant number of projects that still have a `NULL` `project_type` column _and_ no `"level"` field.

```
mysql> select count(id) from storage_apps where id > 52387852 and project_type is null and not INSTR(value, '"level":"');
+-----------+
| count(id) |
+-----------+
|      1504 |
+-----------+
1 row in set (3.93 sec)
```

I took a closer look and these include projects created today, so this is current data.  The projects I looked at are both "hidden" and "standalone" and I'm not entirely sure what that means.  On a spot-check, I also can't find most of these on S3 at all.  A few of them include `"projectType":null,"migratedToS3":true` and these appear to be pixelation widget records.

So... we're still some distance from having a reliable, general way to look up the project type of a channel, but if we have confidence that this unknown set is _not_ `applab` or `gamelab` this might be sufficient for my purposes.  Any other ideas on ways I should be looking for this information?

Reply:
Nice investigation, Brad. I agree, it seems fine to use a method for determining project type that is unreliable for some projects that are not applab or gamelab. No, I can't think of any other existing ways to determine project type besides the ones we've discussed.